File: implementations/spring-boot-starter/src/test/kotlin/io/github/mtempty/ashurbanipal/TableListingPrivilegesTest.kt
package io.github.mtempty.ashurbanipal

import com.zaxxer.hikari.HikariConfig
import com.zaxxer.hikari.HikariDataSource
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.net.URI
import java.sql.Connection
import java.sql.DriverManager
import java.sql.SQLException
import java.sql.SQLFeatureNotSupportedException
import javax.sql.DataSource

private const val SCHEMA = "ashb_test_table_privileges"
private const val ROLE = "ashb_test_table_privileges_role"

private fun jdbcCredentials(): Triple<String, String, String> {
    val databaseUrl = System.getenv("DATABASE_URL")
        ?: error("DATABASE_URL must be set (the devcontainer sets it automatically)")
    val uri = URI(databaseUrl)
    val (user, password) = (uri.userInfo ?: ":").split(":", limit = 2).let { it[0] to it.getOrElse(1) { "" } }
    return Triple("jdbc:postgresql://${uri.host}:${uri.port}${uri.path}", user, password)
}

/** Every connection runs as [ROLE] — USAGE on [SCHEMA] but SELECT on only one of its tables. */
private class SetRoleDataSource(
    private val jdbcUrl: String,
    private val user: String,
    private val password: String,
) : DataSource {
    override fun getConnection(): Connection {
        val conn = DriverManager.getConnection(jdbcUrl, user, password)
        conn.createStatement().use { it.execute("set role $ROLE") }
        return conn
    }

    override fun getConnection(username: String?, password: String?): Connection = getConnection()
    override fun getLogWriter(): java.io.PrintWriter? = null
    override fun setLogWriter(out: java.io.PrintWriter?) {}
    override fun setLoginTimeout(seconds: Int) {}
    override fun getLoginTimeout(): Int = 0
    override fun getParentLogger() = throw SQLFeatureNotSupportedException()
    override fun <T : Any?> unwrap(iface: Class<T>?): T = throw SQLException("not a wrapper")
    override fun isWrapperFor(iface: Class<*>?): Boolean = false
}

/**
 * The table listing, the counts, and the `table` allow-list must all
 * exclude tables the connected role can't `SELECT` (spec/protocol.md
 * §5.2), and an `INSERT`-only table must come back as a clean
 * `NotAllowedException`, never a raw `permission denied` 500 — Kotlin
 * equivalent of `implementations/rust/axum/tests/table_listing_privileges.rs`.
 */
class TableListingPrivilegesTest {
    @Test
    fun `listing and allow-list exclude non-selectable tables`() {
        val (jdbcUrl, user, password) = jdbcCredentials()

        HikariDataSource(
            HikariConfig().apply {
                this.jdbcUrl = jdbcUrl
                username = user
                this.password = password
                maximumPoolSize = 1
                poolName = "table-privileges-setup"
            },
        ).use { setupDs ->
            setupDs.connection.use { conn ->
                conn.createStatement().use { st ->
                    st.execute("drop schema if exists $SCHEMA cascade")
                    st.execute("drop role if exists $ROLE")
                    st.execute("create role $ROLE nosuperuser")
                    // Lets this session `set role` to it in SetRoleDataSource.
                    st.execute("grant $ROLE to current_user")
                    st.execute("create schema $SCHEMA")
                    st.execute("grant usage on schema $SCHEMA to $ROLE")
                    st.execute("create table $SCHEMA.readable (id int primary key, name text)")
                    st.execute("insert into $SCHEMA.readable values (1, 'a'), (2, 'b')")
                    st.execute("create table $SCHEMA.write_only (id int primary key)")
                    st.execute("create table $SCHEMA.no_grant (id int primary key)")
                    st.execute("grant select on $SCHEMA.readable to $ROLE")
                    st.execute("grant insert on $SCHEMA.write_only to $ROLE")
                }
            }

            try {
                val source = PostgresSource(SetRoleDataSource(jdbcUrl, user, password), 5, FilterValidator())
                val opts = QueryOpts(limit = 10, offset = 0, sort = null, descending = false, filter = null)

                assertEquals(
                    listOf("readable"),
                    source.listTables(SCHEMA).map { it.name },
                    "listTables must omit write_only (INSERT only) and no_grant (no privilege)",
                )
                assertEquals(
                    listOf("readable"),
                    source.tableCounts(SCHEMA).map { it.table },
                    "tableCounts must track the same set as listTables",
                )

                source.queryTable(SCHEMA, "readable", opts)

                assertThrows<NotAllowedException>("an INSERT-only table must be rejected, not reach a permission-denied 500") {
                    source.queryTable(SCHEMA, "write_only", opts)
                }
                assertThrows<NotAllowedException>("a table the role has no privilege on must be rejected") {
                    source.queryTable(SCHEMA, "no_grant", opts)
                }
            } finally {
                setupDs.connection.use { conn ->
                    conn.createStatement().use { st ->
                        st.execute("drop schema if exists $SCHEMA cascade")
                        st.execute("drop role if exists $ROLE")
                    }
                }
            }
        }
    }
}

File: implementations/go-nethttp/table_listing_privileges_test.go
package ashurbanipal

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"os"
	"testing"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgconn"
	"github.com/jackc/pgx/v5/stdlib"
)

const (
	tablePrivSchema = "ashb_test_table_privileges"
	tablePrivRole   = "ashb_test_table_privileges_role"
)

// The table listing, the counts, and the `table` allow-list must all
// Excludes non-selectable tables and maps residual SELECT denial to NotAllowed
// (spec/protocol.md §5.2).
func TestListingAndAllowListExcludeNonSelectableTables(t *testing.T) {
	databaseURL := os.Getenv("DATABASE_URL")
	if databaseURL == "" {
		t.Skip("DATABASE_URL not set (the devcontainer sets it automatically)")
	}

	setupDB, err := sql.Open("pgx", databaseURL)
	if err != nil {
		t.Fatalf("opening setup connection: %v", err)
	}
	defer setupDB.Close()

	for _, stmt := range []string{
		fmt.Sprintf("drop schema if exists %s cascade", tablePrivSchema),
		fmt.Sprintf("drop role if exists %s", tablePrivRole),
		fmt.Sprintf("create role %s nosuperuser", tablePrivRole),
		// Lets the pool's sessions `set role` to it in AfterConnect below.
		fmt.Sprintf("grant %s to current_user", tablePrivRole),
		fmt.Sprintf("create schema %s", tablePrivSchema),
		fmt.Sprintf("grant usage on schema %s to %s", tablePrivSchema, tablePrivRole),
		fmt.Sprintf("create table %s.readable (id int primary key, name text)", tablePrivSchema),
		fmt.Sprintf("insert into %s.readable values (1, 'a'), (2, 'b')", tablePrivSchema),
		fmt.Sprintf("create table %s.write_only (id int primary key)", tablePrivSchema),
		fmt.Sprintf("create table %s.no_grant (id int primary key)", tablePrivSchema),
		fmt.Sprintf("grant select on %s.readable to %s", tablePrivSchema, tablePrivRole),
		fmt.Sprintf("grant insert on %s.write_only to %s", tablePrivSchema, tablePrivRole),
	} {
		if _, err := setupDB.Exec(stmt); err != nil {
			t.Fatalf("setup exec %q: %v", stmt, err)
		}
	}
	t.Cleanup(func() {
		setupDB.Exec(fmt.Sprintf("drop schema if exists %s cascade", tablePrivSchema))
		setupDB.Exec(fmt.Sprintf("drop role if exists %s", tablePrivRole))
	})

	connConfig, err := pgx.ParseConfig(databaseURL)
	if err != nil {
		t.Fatalf("parsing DATABASE_URL: %v", err)
	}
	connConfig.AfterConnect = func(ctx context.Context, pc *pgconn.PgConn) error {
		_, err := pc.Exec(ctx, fmt.Sprintf("set role %s", tablePrivRole)).ReadAll()
		return err
	}
	connStr := stdlib.RegisterConnConfig(connConfig)
	defer stdlib.UnregisterConnConfig(connStr)

	db, err := sql.Open("pgx", connStr)
	if err != nil {
		t.Fatalf("opening under-privileged pool: %v", err)
	}
	defer db.Close()

	source := NewPostgresSource(db, 5)
	ctx := context.Background()
	schema := tablePrivSchema
	opts := QueryOpts{Limit: 10, Offset: 0}

	tables, err := source.ListTables(ctx, &schema)
	if err != nil {
		t.Fatalf("ListTables: %v", err)
	}
	if len(tables) != 1 || tables[0].Name != "readable" {
		t.Fatalf("ListTables must omit write_only (INSERT only) and no_grant (no privilege), got %+v", tables)
	}

	counts, err := source.TableCounts(ctx, &schema)
	if err != nil {
		t.Fatalf("TableCounts: %v", err)
	}
	if len(counts) != 1 || counts[0].Table != "readable" {
		t.Fatalf("TableCounts must track the same set as ListTables, got %+v", counts)
	}

	if _, err := source.QueryTable(ctx, &schema, "readable", opts); err != nil {
		t.Fatalf("readable is SELECT-able: %v", err)
	}

	var notAllowed *NotAllowedError
	if _, err := source.QueryTable(ctx, &schema, "write_only", opts); !errors.As(err, &notAllowed) {
		t.Fatalf("an INSERT-only table must be rejected as NotAllowedError, not a permission-denied 500; got %v", err)
	}
	if _, err := source.QueryTable(ctx, &schema, "no_grant", opts); !errors.As(err, &notAllowed) {
		t.Fatalf("a table the role has no privilege on must be rejected as NotAllowedError; got %v", err)
	}
}
